src/Platform/Microsoft.Testing.Platform/Hosts/TestHostBuilder.cs

```diff
@@ -236,8 +237,30 @@ public async Task<ITestHost> BuildAsync(
             Logging.AddProvider((_, _) => loggingState.FileLoggerProvider);
         }
 
-        // Register the server mode log forwarder if needed. We follow the console --diagnostic behavior.
         ICommandLineOptions commandLineOptions = serviceProvider.GetCommandLineOptions();
+
+        // setting the timeout
+        if (commandLineOptions.IsOptionSet(PlatformCommandLineProvider.TimeoutOptionKey) && commandLineOptions.TryGetOptionArgumentList(PlatformCommandLineProvider.TimeoutOptionKey, out string[]? args))
+        {
+            string arg = args[0];
+            int size = arg.Length;
+            if (!float.TryParse(arg[..(size - 1)], out float value))
+            {
+                throw new InvalidOperationException(PlatformResources.PlatformCommandLineTimeoutArgumentErrorMessage);
+            }
+
+            TimeSpan timeout = char.ToLowerInvariant(arg[size - 1]) switch
+            {
+                'h' => TimeSpan.FromHours(value),
+                'm' => TimeSpan.FromMinutes(value),
+                's' => TimeSpan.FromSeconds(value),
+                _ => throw new InvalidOperationException(PlatformResources.PlatformCommandLineTimeoutArgumentErrorMessage),
+            };
+
+            testApplicationCancellationTokenSource.CancelAfter(timeout);
```

Comment:
I think that's fine as-is but I'd like to get @MarcoRossignoli's view. Marco, shall we do that directly when initializing the test app source object?

Reply:
I think it's fine we build the appct before the command line.